### PR TITLE
Introduce exclusive search

### PR DESCRIPTION
**Pull request by antarcticrainforest**
As we discussed earlier. The restAPI is not able to perform a search the should NOT include certain values. This addresses this.

The old API realised an exclusive search by using `<facet_name>_not_=value`. I figured that it is more straightforward and allow for a more fine-grained search if we used normal lucene syntax. That is `facet_name=-value` or `facet_name=not value` or `facet_name=!value`. That way we don't have to deal with the facet checker.

If on the other hand, anyone says we should have the old way `<facet_name>_not_=value` please shout now.

### CODE REVIEW

File: freva-rest/src/databrowser_api/core.py
Comment (eelucio) [inline freva-rest/src/databrowser_api/core.py:777]: should not be
```python
                negative.append(search_value.lstrip("NOT ").lstrip("not "))
```
instead ?

File: freva-rest/src/databrowser_api/core.py
Comment (eelucio) [inline freva-rest/src/databrowser_api/core.py:775]: what happens with this case, is it lost? Could this be:
```python
                positive.append(search_value.lower())
```

instead?

File: freva-rest/src/databrowser_api/core.py
Comment (eelucio) [inline freva-rest/src/databrowser_api/core.py:802]: I do not get it, if the negation is done in https://github.com/FREVA-CLINT/freva-nextgen/blob/668fcc086d72322d3047254819266b18f43d2025/freva-rest/src/databrowser_api/core.py#L777
what happens if you do it again here
```python
            if query_neg:
                query.append(f"-{key.lower()}:({query_neg})")
```

I added to query_neg so it would strip either not, NOT, ! or - 
and then add that minus to the query itself

File: docs/source/databrowser/cli.rst
Comment (eelucio) [inline docs/source/databrowser/cli.rst:114]: sorry for this dumb question, but has been the cli (freva core) changed so that is reflected? or does it go independently?

**Review by eelucio — CHANGES_REQUESTED** [review]
I may be wrong but I think there is one small issue there.

on the other hand I do not think the syntax change in the cli will be breaking much (I would pressume that this is a fairly unused capability)

File: docs/source/databrowser/cli.rst
Comment (MoSHad91) [reply docs/source/databrowser/cli.rst:111]: and yes, it seems it covers curly braces for neg as well: https://github.com/FREVA-CLINT/freva-nextgen/blob/contains-not/freva-rest/src/databrowser_api/core.py#L786

**Review by MoSHad91 — APPROVED** [review]
this is one just a comment on the already developed part which has not been documented ... lgtm🚀

File: freva-rest/src/databrowser_api/core.py
Comment (antarcticrainforest) [reply freva-rest/src/databrowser_api/core.py:775]: We don't need that. because `search_value` is already transformed to lowercase when the search key is not file or uri. We do that because files/uri stored without making the lowercase (we need to access them)

File: freva-rest/src/databrowser_api/core.py
Comment (antarcticrainforest) [reply freva-rest/src/databrowser_api/core.py:777]: Yes, I know the logic is kind of awkward. The idea is to replace leading *not*'s by leading -. I guess I could aswell do it like this:

```python
if search_value.lower().startswith("not "):
    negative.append(search_value.lstrip("NOT ").lstrip("not "))
elif search_value[0] in ("!", "-"):
     negative.append(search_value.lstrip("!").lstrip("-"))
else:
     positive.append(search_value)

```

File: freva-rest/src/databrowser_api/core.py
Comment (antarcticrainforest) [reply freva-rest/src/databrowser_api/core.py:802]: We are not doing it again. My idea was to collect negative search statements (searches that should NOT contain the search result) and positive ones (facets that should contain search results). Once we have those we need to translate this to lucene syntax. Unfortunately, this doesn't work in lucene:

`{"fq": "project":"-cmip6"} ` it needs to be `{"fq":"-project":"cmip6"}`.

I am not super happy about this but there is nothing we can do.

In my opinion, something like project=-cmip6 is easy to understand and it is easier to implement than putting the not condition to the search key, because this: -project=cmip6 doesn't work in python nor at the cli. I guess this is why the old system had the` _not_ `construct. project_not_ = cmip.


We can reintroduce this for the sake of backward compatibility. But then we'd have to change the facet checker. Might not be a big deal though? Any opnios @MoSHad91 @eelucio

File: docs/source/databrowser/cli.rst
Comment (antarcticrainforest) [reply docs/source/databrowser/cli.rst:114]: That's a very good question. No, it hasn't yet. I think we should switch to using the databrowser api also in the old freva core.

**Comment by antarcticrainforest** [comment]
I've also added the `facet_not_=value` construct back to the code. So there is nothing to be done from the plugin side.

The side effect it that it doesn't matter where the `_not_` is located so `_not_facet=value` or `facet_not_=value` will have the same effect even `fa_not_cet=value` if you think that is problematic let me know.

**Comment by eelucio** [comment]
LGTM!

One last dumb minor question:

Is there any reason why you define a function [`foreward_lookup()`](https://github.com/FREVA-CLINT/freva-nextgen/blob/contains-not/freva-rest/src/databrowser_api/core.py#L263) instead of e.g. `forward_lookup()`

If I want to do some minor typo correction should I open a new branch once this is merge, or wait untill all is merge?

**Comment by eelucio** [comment]
> I've also added the `facet_not_=value` construct back to the code. So there is nothing to be done from the plugin side.
> 
> The side effect is, that it doesn't matter where the `_not_` is located so `_not_facet=value` or `facet_not_=value` will have the same effect even `fa_not_cet=value` if you think that is problematic let me know.

The fact that we have some facets written with underscore e.g. [`file_no_version` in the managed schema](https://github.com/FREVA-CLINT/freva-service-config/blob/main/solr/managed-schema.xml#L65) should not be a problem, right?

I guess that for the future we will just need to be sure that nobody enters a facet that has a `_not_` in between the name
as the above mentioned param is dangerously close to this statement. But as far as I know it was not searchable, right?


otherwise LGTM!

one (very) minor comment:
is there any reason why there is a function called [`foreward_lookup()`](https://github.com/FREVA-CLINT/freva-nextgen/blob/contains-not/freva-rest/src/databrowser_api/core.py#L263) (old english) instead of `forward_lookup()`? I hope the typo correctors do not change it automatically 

and one last one, shall I'd like to correct some typos (e.g. in docstrings etc.) should I wait until many other branches are merged?

**Pull request merged**
